feat(unleash): Add unleash feature flagging

damassi opened the pull request:
This adds Unleash feature flag support to Metaphysics, via Unleash's [node-client](https://github.com/Unleash/unleash-client-node). 

Also adds our first feature flag for globally toggling off blurhash.

Right now the primary API is:
```tsx
import { isFeatureFlagEnabled } from "lib/featureFlags"

const isBlurhashEnabled = isFeatureFlagEnabled(
  "diamond_blurhash-enabled-globally"
)
```

An optional [unleash context](https://docs.getunleash.io/reference/unleash-context) can be passed in via a second argument: 

```tsx
isFeatureFlagEnabled(flagName, context)
```

mzikherman commented on docs/feature-flags.md:
```diff
@@ -0,0 +1,31 @@
+# Feature Flags via Unleash
+
+We manage feature toggles via [Unleash](unleash.artsy.net), which can be managed via [tools.artsy.net](https://tools.artsy.net/feature-flags).
```
(needs to be fully qualified URL to unleash)

mzikherman commented on src/lib/featureFlags.ts:
```diff
@@ -0,0 +1,52 @@
+  return unleashClient.isEnabled(flag)
+}
+
+export function getUnleashClient() {
```
This getter doesn't look to be used.

damassi replied:
removed for now 👍

damassi commented on src/config.ts:
```diff
@@ -76,6 +76,9 @@ const {
   SYSTEM_ENVIRONMENT,
   UNLEASH_ADMIN_TOKEN,
+  UNLEASH_API,
+  UNLEASH_APP_NAME,
```
@joeyAghion / @mc-jones - to confirm, this is just a raw string, correct? Its not something to configure in admin. Noticing we have `force-staging`, etc. I added here `metaphysics-staging`.

mc-jones replied:
Yea, I believe that is correct. Unleash should pick up the app name automatically, and expose some [details](https://unleash.artsy.net/applications/metaphysics-staging) about it in the UI. 

The more important config is the `UNLEASH_SERVER_KEY`.  I believe there needs to be two seperate keys. One set to `development` (staging and local) and one set to `production`. IIRC this is how Unleash allows you to toggle the feature on and off in different environments.

damassi replied:
Cool, i've updated this for staging and prod, via hokusai, so we should be good 👍

mzikherman reviewed (approved):
Nice! Looks pretty simple. Initializing Unleash async (even if some initial requests when a new pod spins up might see 'false' for all flags before it syncs with the Unleash server) seems the right thing to do.

joeyAghion reviewed (commented):
It's important to note somewhere that, since we're not providing any Unleash "context" to these calls, some usual strategies for determining toggle state (like user id-based) will have no effect. This will only work for toggles that are on or off for the whole environment.

Which reminds me: shouldn't there be an "environment" somewhere in this initialization? Or is that just decided by the configured key?

mc-jones commented on src/lib/featureFlags.ts:
```diff
@@ -0,0 +1,43 @@
+    url: UNLEASH_API,
+    appName: UNLEASH_APP_NAME,
+    customHeaders: { Authorization: UNLEASH_SERVER_KEY },
+  })
```
Do we run the risk of instantiating multiple clients unless we guard against it? If `src/index.ts` is run more than once, than I think this might be an issue.

damassi replied:
Index.ts isn't run more than once (unless in dev mode, and we've hot reloaded a change), so all ok!

damassi commented:
Ah, nice catch @joeyAghion. Added a second argument slot and updated the docs. 

Re environment, it comes from the server key; eg, 
```
UNLEASH_SERVER_KEY=*:development.....
```

damassi commented:
Aight, gonna merge this and get it out to staging 👍 Thanks all for the review
